shell/platform/windows/win32_task_runner.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/win32_task_runner.h"
+
+#include <atomic>
+#include <utility>
+
+namespace flutter {
+
+Win32TaskRunner::Win32TaskRunner(std::thread::id main_thread_id,
+                                 TaskExpiredCallback on_task_expired)
+    : main_thread_id_(main_thread_id),
+      on_task_expired_(std::move(on_task_expired)),
+      wait_duration_(std::chrono::seconds::max()) {}
+
+Win32TaskRunner::~Win32TaskRunner() = default;
+
+bool Win32TaskRunner::RunsTasksOnCurrentThread() const {
+  return std::this_thread::get_id() == main_thread_id_;
+}
+
+void Win32TaskRunner::GetMessage(LPMSG message,
+                                 HWND window,
+                                 UINT message_filter_min,
+                                 UINT message_filter_max) {
+  BOOL got_message = FALSE;
+
+  // Repeat until a message is retrieved from the win32 message queue.
+  while (!got_message) {
+    const auto now = TaskTimePoint::clock::now();
+
+    // Wait for either a window message or timeout.
+    // if a timeout has been set at the end of a previous iteration, wait for
+    // that duration, else wait until there a new message is posted.
+    MsgWaitForMultipleObjects(
+        0, NULL, FALSE, (DWORD)(wait_duration_.count() * 1000), QS_ALLEVENTS);
```

Comment:
No C-style casts.

Also, won't this overflow? Why not just make wait_duration_ milliseconds if that's the unit it's going to be consumed in?